flatland_plugins/src/laser.cpp

```diff
@@ -162,64 +166,108 @@ void Laser::ComputeLaserRanges() {
   // Conver to Box2D data types
   b2Vec2 laser_origin_point(v_world_laser_origin_(0), v_world_laser_origin_(1));
 
-  // Results vector
-  std::vector<std::future<std::pair<float, float>>> results(
-      laser_scan_.ranges.size());
-
-  // loop through the laser points and call the Box2D world raycast by
-  // enqueueing the callback
-  for (unsigned int i = 0; i < laser_scan_.ranges.size(); ++i) {
-    results[i] =
-        pool_.enqueue([i, this, laser_origin_point] {  // Lambda function
-          b2Vec2 laser_point;
-          laser_point.x = m_world_laser_points_(0, i);
-          laser_point.y = m_world_laser_points_(1, i);
-          LaserCallback cb(this);
-
-          GetModel()->GetPhysicsWorld()->RayCast(&cb, laser_origin_point,
-                                                 laser_point);
-
-          if (!cb.did_hit_) {
-            return std::make_pair<float, float>(NAN, 0);
-          } else {
-            return std::make_pair<float, float>(
-                cb.fraction_ * this->range_ + this->noise_gen_(this->rng_),
-                static_cast<float>(cb.intensity_));
-          }
-        });
+  // Results vector : clustered to avoid too many locks
+  constexpr size_t clusterSize = 10;
+  // The last cluster may have a different size
+  size_t lastClusterSize = laser_scan_.ranges.size() % clusterSize;
+  size_t nbCluster =
+      laser_scan_.ranges.size() / clusterSize + (lastClusterSize == 0 ? 0 : 1);
+  if (lastClusterSize == 0) lastClusterSize = clusterSize;
+  std::vector<std::future<std::vector<std::pair<float, float>>>> results(
```

Comment:
Can this data structure be reused? In principle, a raycast just needs to write to a specific memory location.

Reply:
I don't think so, the purpose is only to split the raycast calls into buckets. The pool thread dispatches the tasks with the intermediate use of package tasks, so it returns only one future when enqueing each task. This future acts as a flag indicating that the whole bucket is ready. It could be indeed possible to write in a predefined vector or other memory location, by implementing a index logic to safely write in this memory.

Reply:
Yes, let's leave this for a future PR